### ISSUE TITLE
[By Nov 1, 2021] Update targetSdkversion to 30

### ISSUE THREAD

**Issue by lognaturel** [getodk/collect#4533]
Must have for any release that will go out after Nov 1, 2021.

I have done a quick read through https://developer.android.com/about/versions/11/behavior-changes-11 and I believe the changes needed are:
- [x] Remove all legacy storage access
- ~~Change `READ_PHONE_STATE` to only API 29 and under, [add `READ_PHONE_NUMBERS`](https://developer.android.com/about/versions/11/privacy/permissions#phone-numbers)~~
- [x] Look into [APK Scheme Signature v2](https://developer.android.com/about/versions/11/behavior-changes-11#minimum-signature-scheme). I believe we need to change the release process.
- ~~Consider any action to take given that [permissions won't be requested after being repeatedly denied](https://developer.android.com/about/versions/11/privacy/permissions#dialog-visibility).~~

**Comment by seadowg** [comment]
It doesn't actually look like there is any advantage to us using `READ_PHONE_NUMBERS` over `READ_PHONE_STATE` - the OS still presents the annoyingly scary "allow ODK Collect to make and manage phone calls" text. It's non-trivial to implement as well, as we'd need to add some logic to `PermissionsProvider` to switch on different SDK levels (which means tests). I'll pull that out as a separate issue as we still should do it eventually (I imagine `READ_PHONE_STATE` will go away or become harder to get) but I don't think we should do it here.

**Comment by seadowg** [comment]
I verified our most recent proper release APK (v2021.2.4) and it looks like we're already using the v2 Signature. Using this:

```bash
apksigner verify --print-certs --min-sdk-version 21 -v <apk path>
```

I got:

```bash
Verified using v1 scheme (JAR signing): true
Verified using v2 scheme (APK Signature Scheme v2): true
Verified using v3 scheme (APK Signature Scheme v3): false
Verified using v4 scheme (APK Signature Scheme v4): false
Verified for SourceStamp: false
```

I'm guessing v2 is used by default in the version of the Gradle tools we're using.

**Comment by seadowg** [comment]
Looking at the "permissions won't be requested after being repeatedly denied" thing it does seem like we need to make a change. Currently, if the user denies a permission twice then they'll see something like this every time they encounter that permission again:

```
Without this permission Collect can't access the microphone. Please try again when you are ready to grant it.
```

Given they don't see the permission request again, I think it'd be very unclear what action they should take (unless they're an Android power user and know to go to Settings > App > ODK Collect > Permissions). It looks like a lot of apps (see Google Maps for an example) are handling this by adding an "Open Settings" button to the dialog/snackbar they pop up on denial that opens the App Details page in System Settings (the "ODK Collect" part of the navigation stack example). I think we should also do that or I'd be worried we'd leave new users in an awkward state where they've denied permissions and don't know how to grant them. @yanokwa thoughts?

**Comment by yanokwa** [comment]
I think following what other apps are doing makes sense to me. And maybe we add some copy that explains what they should do when the hit Open Settings. Any concerns/downsides?

**Comment by seadowg** [comment]
> And maybe we add some copy that explains what they should do when the hit Open Settings.

Yeah, good point. I'll update the copy for the messages as well, so it fits better with that flow.

> Any concerns/downsides?

I suspect there will still be a few people who will click through to settings and be intimidated by the App Details page, but that's the best Android's going to give us, I think. It might be that in future we want to think about having permission requests as part of an onboarding flow to allow them to be setup at the same time as projects are added etc. I think we'll have to listen out for feedback on this as people upgrade to Android 11.

**Comment by seadowg** [comment]
@yanokwa actually, after testing this on a different OS level and [re-reading the docs](https://developer.android.com/about/versions/11/privacy/permissions#dialog-visibility), I realized that @lognaturel and I had both made the mistake thinking this was a "targeting Android 11" change rather than a plain ol' operating system change. By this, I mean that we run into this problem with existing versions of the app on Android 11 already (currently just over 10% of our monthly users). I'll pull what we've discussed out to a separate issue.
